File: src/main/java/com/example/view_shoppingmall/service/ProductService.java
package com.example.view_shoppingmall.service;

import com.example.view_shoppingmall.dto.ProductDTO;
import com.example.view_shoppingmall.entity.Product;
import com.example.view_shoppingmall.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ProductService {

    private final ProductRepository productRepository;

    // 모든 상품 조회
    public List<ProductDTO> getAllProducts() {
        return productRepository.findAll().stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    // 상품 ID로 조회
    public ProductDTO getProductById(Long id) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("상품을 찾을 수 없습니다. ID: " + id));
        return convertToDTO(product);
    }

    // 카테고리별 상품 조회
    public List<ProductDTO> getProductsByCategory(String category) {
        return productRepository.findByCategory(category).stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    // 상품명 검색
    public List<ProductDTO> searchProducts(String keyword) {
        return productRepository.findByNameContaining(keyword).stream()
                .map(this::convertToDTO)
                .collect(Collectors.toList());
    }

    // 상품 등록
    @Transactional
    public ProductDTO createProduct(ProductDTO productDTO) {
        Product product = convertToEntity(productDTO);
        Product savedProduct = productRepository.save(product);
        return convertToDTO(savedProduct);
    }

    // 상품 수정
    @Transactional
    public ProductDTO updateProduct(Long id, ProductDTO productDTO) {
        Product product = productRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("상품을 찾을 수 없습니다. ID: " + id));

        product.setName(productDTO.getName());
        product.setDescription(productDTO.getDescription());
        product.setPrice(productDTO.getPrice());
        product.setStock(productDTO.getStock());
        product.setImageUrl(productDTO.getImageUrl());
        product.setCategory(productDTO.getCategory());

        Product updatedProduct = productRepository.save(product);
        return convertToDTO(updatedProduct);
    }

    // 상품 삭제
    @Transactional
    public void deleteProduct(Long id) {
        if (!productRepository.existsById(id)) {
            throw new RuntimeException("상품을 찾을 수 없습니다. ID: " + id);
        }
        productRepository.deleteById(id);
    }

    // Entity -> DTO 변환
    private ProductDTO convertToDTO(Product product) {
        return new ProductDTO(
                product.getId(),
                product.getName(),
                product.getDescription(),
                product.getPrice(),
                product.getStock(),
                product.getImageUrl(),
                product.getCategory());
    }

    // DTO -> Entity 변환
    private Product convertToEntity(ProductDTO dto) {
        Product product = new Product();
        product.setName(dto.getName());
        product.setDescription(dto.getDescription());
        product.setPrice(dto.getPrice());
        product.setStock(dto.getStock());
        product.setImageUrl(dto.getImageUrl());
        product.setCategory(dto.getCategory());
        return product;
    }
}

File: frontend/src/services/productService.js
import apiClient from './api';

/**
 * 상품 관련 API 호출 함수들
 */

// 모든 상품 조회
export const getAllProducts = async () => {
    try {
        const response = await apiClient.get('/products');
        return response.data;
    } catch (error) {
        console.error('상품 목록 조회 실패:', error);
        throw error;
    }
};

// 특정 상품 조회
export const getProductById = async (id) => {
    try {
        const response = await apiClient.get(`/products/${id}`);
        return response.data;
    } catch (error) {
        console.error('상품 조회 실패:', error);
        throw error;
    }
};

// 카테고리별 상품 조회
export const getProductsByCategory = async (category) => {
    try {
        const response = await apiClient.get(`/products/category/${category}`);
        return response.data;
    } catch (error) {
        console.error('카테고리별 상품 조회 실패:', error);
        throw error;
    }
};

// 상품 검색
export const searchProducts = async (keyword) => {
    try {
        const response = await apiClient.get('/products/search', {
            params: { keyword }
        });
        return response.data;
    } catch (error) {
        console.error('상품 검색 실패:', error);
        throw error;
    }
};

// 상품 등록
export const createProduct = async (productData) => {
    try {
        const response = await apiClient.post('/products', productData);
        return response.data;
    } catch (error) {
        console.error('상품 등록 실패:', error);
        throw error;
    }
};

// 상품 수정
export const updateProduct = async (id, productData) => {
    try {
        const response = await apiClient.put(`/products/${id}`, productData);
        return response.data;
    } catch (error) {
        console.error('상품 수정 실패:', error);
        throw error;
    }
};

// 상품 삭제
export const deleteProduct = async (id) => {
    try {
        await apiClient.delete(`/products/${id}`);
    } catch (error) {
        console.error('상품 삭제 실패:', error);
        throw error;
    }
};
